src/Miunie.Discord/Providers/HelpCommandProvider.cs

```diff
@@ -0,0 +1,129 @@
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  Miunie is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with Miunie. If not, see <https://www.gnu.org/licenses/>.
+
+using Discord.Commands;
+using Discord.WebSocket;
+using Miunie.Core.Entities;
+using Miunie.Core.Providers;
+using Miunie.Discord.Attributes;
+using Miunie.Discord.Embeds;
+using Miunie.Discord.Entities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Miunie.Discord
+{
+    public class HelpCommandProvider
+    {
+        private readonly CommandService _commandService;
+        private readonly ILanguageProvider _lang;
+
+        public HelpCommandProvider(CommandService commandService, ILanguageProvider lang)
+        {
+            _commandService = commandService;
+            _lang = lang;
+        }
+
+        public HelpResult GetDefault()
+            => new HelpResult
+            {
+                Title = _lang.GetPhrase(PhraseKey.USER_EMBED_HELP_TITLE.ToString()),
+                Sections = GetAllModuleSections()
+            };
+
+        public HelpResult Search(string input)
+        {
+            var sections = new List<HelpSection>();
+            foreach (CommandInfo command in GetCommands(input))
+            {
+                sections.Add(GetSection(command));
+            }
+
+            return new HelpResult()
+            {
+                Sections = sections
+            };
+        }
+
+        public async Task ShowDefaultHelpAsync(ISocketMessageChannel channel)
+            => await channel.SendMessageAsync(embed: EmbedConstructor.CreateHelpEmbed(GetDefault()));
+
+        public async Task ShowCommandHelpAsync(ISocketMessageChannel channel, string input)
+            => await channel.SendMessageAsync(embed: EmbedConstructor.CreateHelpEmbed(Search(input)));
+
+        private IEnumerable<CommandInfo> GetCommands(string input)
+        {
+            SearchResult result = _commandService.Search(input);
+
+            if (!result.IsSuccess)
+            {
+                throw new Exception(result.ErrorReason);
+            }
+
+            return result.Commands.Select(x => x.Command);
+        }
+
+        private IEnumerable<HelpSection> GetAllModuleSections()
+        {
+            var sections = new List<HelpSection>();
+
+            foreach (ModuleInfo module in _commandService.Modules)
+            {
+                sections.Add(GetSection(module));
+            }
+
+            return sections;
```

Comment:
I think this can just be:

```cs
return _commandService.Modules.Select(GetSection);
```

instead of the entire method. And honesty, at that point we don't even need this method to exist.

We could just use the suggested one-liner where `GetAllModuleSelections` is used. 🙂